src/transport_ofi.h

```diff
@@ -658,6 +658,139 @@ void shmem_transport_put_nb(shmem_transport_ctx_t* ctx, void *target, const void
+                                  .msg_iov = &msg_iov,
+                                  .desc = NULL,
+                                  .iov_count = 1,
+                                  .addr = GET_DEST(dst),
+                                  .rma_iov = &rma_iov,
+                                  .rma_iov_count = 1,
+                                  .context = frag_source,
+                                  .data = 0
+                                };
+
+        SHMEM_TRANSPORT_OFI_CTX_LOCK(ctx);
+        while (frag_source < (((uint8_t *) source) + len)) {
+            frag_len = MIN(shmem_transport_ofi_max_msg_size, 
+                          (size_t) (((uint8_t *) source) + len - frag_source));
+            polled = 0;
+
+            msg_iov.iov_base = frag_source;
+            msg_iov.iov_len = frag_len;
+
+            rma_iov.addr = frag_target;
+            rma_iov.len = frag_len;
+
+            msg.msg_iov = &msg_iov;
+            msg.rma_iov = &rma_iov;
+            msg.context = frag_source;
+
+            SHMEM_TRANSPORT_OFI_CNTR_INC(&ctx->pending_put_cntr);
+
+            do {
+                ret = fi_writemsg(ctx->ep, &msg, FI_DELIVERY_COMPLETE);
+            } while (try_again(ctx, ret, &polled));
+
+            frag_source += frag_len;
+            frag_target += frag_len;
+        }
+        SHMEM_TRANSPORT_OFI_CTX_UNLOCK(ctx);
+    }
+
+    /* Transmit the signal */
+    shmem_transport_ofi_get_mr(sig_addr, pe, &addr, &key);
+    polled = 0;
+    ret = 0;
+
+    SHMEM_TRANSPORT_OFI_CTX_LOCK(ctx);
+    SHMEM_TRANSPORT_OFI_CNTR_INC(&ctx->pending_put_cntr);
+
+    const struct fi_ioc msg_iov_signal = {
+                                          .addr = (uint8_t *) &signal,
+                                          .count = 1
+                                        };
+    const struct fi_rma_ioc rma_iov_signal = {
+                                               .addr = (uint64_t) addr,
+                                               .count = 1,
+                                               .key = key
+                                             };
+    const struct fi_msg_atomic msg_signal = {
+                                           .msg_iov = &msg_iov_signal,
+                                           .desc = NULL,
+                                           .iov_count = 1,
+                                           .addr = GET_DEST(dst),
+                                           .rma_iov = &rma_iov_signal,
+                                           .rma_iov_count = 1,
+                                           .datatype = FI_UINT64,
+                                           .op = FI_ATOMIC_WRITE,
+                                           .context = (uint8_t *) &signal,
+                                           .data = 0
+                                         };
+
+    do {
+        ret = fi_atomicmsg(ctx->ep, &msg_signal, FI_DELIVERY_COMPLETE | FI_FENCE | FI_INJECT);
```

Comment:
The open question is, whether we also need to use an atomic to implement `shmem_p`.  Could you add a comment here and open an issue so we remember to revisit this?

Reply:
Issue opened.